src/main/fc/rc.c

```diff
@@ -524,38 +523,142 @@ static FAST_CODE void processRcSmoothingFilter(void)
         float *dst = i == THROTTLE ? &rcCommand[i] : &setpointRate[i];
         if (rcSmoothingData.filterInitialized) {
-            *dst = pt3FilterApply(&rcSmoothingData.filter[i], rxDataToSmooth[i]);
+            *dst = pt3FilterApply(&rcSmoothingData.filterSetpoint[i], rxDataToSmooth[i]);
         } else {
             // If filter isn't initialized yet, as in smoothing off, use the actual unsmoothed rx channel data
             *dst = rxDataToSmooth[i];
         }
-    }
 
-    // for ANGLE and HORIZON, smooth rcDeflection on pitch and roll to avoid setpoint steps
-    bool smoothingNeeded = (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(HORIZON_MODE)) && rcSmoothingData.filterInitialized;
+    }
     for (int axis = FD_ROLL; axis <= FD_YAW; axis++) {
-        if (smoothingNeeded && axis < FD_YAW) {
-            rcDeflectionSmoothed[axis] = pt3FilterApply(&rcSmoothingData.filterDeflection[axis], rcDeflection[axis]);
+        // Feedforward smoothing
+        feedforwardSmoothed[axis] = pt3FilterApply(&rcSmoothingData.filterFeedforward[axis], feedforwardRaw[axis]);
+        // Horizon mode smoothing of rcDeflection on pitch and roll to provide a smooth angle element
+        const bool smoothRcDeflection = FLIGHT_MODE(HORIZON_MODE) && rcSmoothingData.filterInitialized;
+        if (smoothRcDeflection && axis < FD_YAW) {
+            rcDeflectionSmoothed[axis] = pt3FilterApply(&rcSmoothingData.filterRcDeflection[axis], rcDeflection[axis]);
         } else {
             rcDeflectionSmoothed[axis] = rcDeflection[axis];
         }
     }
-
 }
 #endif // USE_RC_SMOOTHING_FILTER
 
+FAST_CODE_NOINLINE void initAveraging (void)
+{
+    feedforwardAveraging = pidGetFeedforwardAveraging();
+    for (int i = 0; i < XYZ_AXIS_COUNT; i++) {
+        laggedMovingAverageInit(&feedforwardDeltaAvg[i].filter, feedforwardAveraging + 1, (float *)&feedforwardDeltaAvg[i].buf[0]);
+    }
+}
+
+FAST_CODE_NOINLINE void calculateFeedforward (int axis)
+{
+    const float rxInterval = currentRxRefreshRate * 1e-6f; // seconds
+    const float rxRate = 1.0f / rxInterval;
+
+    const float feedforwardSmoothFactor = pidGetFeedforwardSmoothFactor();
+    const float feedforwardJitterFactor = pidGetFeedforwardJitterFactor();
+    const float feedforwardTransitionFactor = pidGetFeedforwardTransitionFactor();
+    const float feedforwardBoostFactor = pidGetFeedforwardBoostFactor();
+
+    if (feedforwardAveraging != pidGetFeedforwardAveraging()) {
+        initAveraging();
+    }
+
+    const float rcCommandDeltaAbs = fabsf(rcCommand[axis] - prevRcCommand[axis]);
+    prevRcCommand[axis] = rcCommand[axis];
+
+    float setpoint = rawSetpoint[axis];
+    float feedforwardDelta = (setpoint - prevSetpoint[axis]);
+    float absPrevSetpointDelta = fabsf(prevSetpointDelta[axis]);
+    float setpointAcceleration = 0.0f;
+
+    // calculate an attenuator from average of two most recent rcCommand deltas vs jitter threshold
+    float ffAttenuator = 1.0f;
```

Comment:
Good catch, that was dumb of me, no clue why I did that.  Will fix.

Reply:
`feedforwardAttenuator` doesn't really tell me more either. That we are in a function called `calculateFeedforward` is enough of a hint that the variable has something to do with feedforward.
Maybe call it `jitterAttenuator` again and use different variables for the other cases? 
It might look inefficient but the code will probably be easier to follow and the compiler will probably sort things out in the end.

Reply:
After checking the code functionally, I needed a value to force some lines to zero, independently of a value to attenuate on the basis of the jitter attenuation code.  So we're back to the original `jitterAttenuator` again :-)